### DISCUSSION TITLE
Cache is becoming invalidated between calls has() and get()

### DISCUSSION THREAD

**Discussion by f1amy** [spatie/laravel-responsecache#342] [General]
Trying to reopen [#137](https://github.com/spatie/laravel-responsecache/issues/137).

For some reason after the cache repository confirms the response [`hasBeenCached()`](https://github.com/spatie/laravel-responsecache/blob/2a9cf083839e5ec5a7cd968e7779c85863901b6e/src/Middlewares/CacheResponse.php#L29),
trying to [`getCachedResponseFor()`](https://github.com/spatie/laravel-responsecache/blob/2a9cf083839e5ec5a7cd968e7779c85863901b6e/src/Middlewares/CacheResponse.php#L32) fails as [`unserialize()`](https://github.com/spatie/laravel-responsecache/blob/2a9cf083839e5ec5a7cd968e7779c85863901b6e/src/ResponseCacheRepository.php#L40) is getting `null` instead of a `string` from the cache.
This happens very rarely.

Exception:
```
Symfony\Component\Debug\Exception\FatalThrowableError: Argument 1 passed to Spatie\ResponseCache\Serializers\DefaultSerializer::unserialize() must be of the type string, null given, called in /srv/data/code/vendor/spatie/laravel-responsecache/src/ResponseCacheRepository.php on line 41
#58 /vendor/spatie/laravel-responsecache/src/Serializers/DefaultSerializer.php(20): Spatie\ResponseCache\Serializers\DefaultSerializer::unserialize
#57 /vendor/spatie/laravel-responsecache/src/ResponseCacheRepository.php(41): Spatie\ResponseCache\ResponseCacheRepository::get
#56 /vendor/spatie/laravel-responsecache/src/ResponseCache.php(74): Spatie\ResponseCache\ResponseCache::getCachedResponseFor
#55 /vendor/spatie/laravel-responsecache/src/Middlewares/CacheResponse.php(33): Spatie\ResponseCache\Middlewares\CacheResponse::handle
#54 /vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(163): Illuminate\Pipeline\Pipeline::Illuminate\Pipeline\{closure}
#53 /vendor/laravel/framework/src/Illuminate/Routing/Pipeline.php(53): Illuminate\Routing\Pipeline::Illuminate\Routing\{closure}
```

**Comment by freekmurze** [comment]
I'd welcome a PR to add a failing test to reproduce this error.

**Comment by f1amy** [comment]
I can do that, but as I said this issue happens rarely, so the test would be:
```
public function it_throws_an_exception_when_response_is_null()
{
    $this->expectException(TypeError::class);

    app(Serializer::class)->unserialize(null);
}
```
And I think that's not a good test, since the fix should be somewhere in [Middlewares/CacheResponse.php](https://github.com/spatie/laravel-responsecache/blob/2a9cf083839e5ec5a7cd968e7779c85863901b6e/src/Middlewares/CacheResponse.php#L32).

**Comment by lisotton** [comment]
I'm also having this issues and as I have hundreds of thousands of requests per day, I have this quite often. 
```
[2021-08-04 17:11:05] production.ERROR: Spatie\\ResponseCache\\Serializers\\DefaultSerializer::unserialize(): Argument #1 ($serializedResponse) must be of type string, null given, called in /var/www/vendor/spatie/laravel-responsecache/src/ResponseCacheRepository.php on line 40 
TypeError(code: 0): Spatie\\\\ResponseCache\\\\Serializers\\\\DefaultSerializer::unserialize(): Argument #1 ($serializedResponse) must be of type string, null given, called in /var/www/vendor/spatie/laravel-responsecache/src/ResponseCacheRepository.php on line 40 at /var/www/vendor/spatie/laravel-responsecache/src/Serializers/DefaultSerializer.php:20)
[stacktrace]
#0 /var/www/vendor/spatie/laravel-responsecache/src/ResponseCacheRepository.php(40): Spatie\\\\ResponseCache\\\\Serializers\\\\DefaultSerializer->unserialize()
#1 /var/www/vendor/spatie/laravel-responsecache/src/ResponseCache.php(72): Spatie\\\\ResponseCache\\\\ResponseCacheRepository->get()
#2 /var/www/vendor/spatie/laravel-responsecache/src/Middlewares/CacheResponse.php(32): Spatie\\\\ResponseCache\\\\ResponseCache->getCachedResponseFor()
#3 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Spatie\\\\ResponseCache\\\\Middlewares\\\\CacheResponse->handle()
#4 /var/www/app/Http/Middleware/BodyTimestamp.php(19): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#5 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): App\\\\Http\\\\Middleware\\\\BodyTimestamp->handle()
#6 /var/www/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#7 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\\\View\\\\Middleware\\\\ShareErrorsFromSession->handle()
#8 /var/www/app/Http/Middleware/StartSession.php(25): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#9 /var/www/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): App\\\\Http\\\\Middleware\\\\StartSession->handleStatefulRequest()
#10 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\\\Session\\\\Middleware\\\\StartSession->handle()
#11 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#12 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(697): Illuminate\\\\Pipeline\\\\Pipeline->then()
#13 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(672): Illuminate\\\\Routing\\\\Router->runRouteWithinStack()
#14 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(636): Illuminate\\\\Routing\\\\Router->runRoute()
#15 /var/www/vendor/laravel/framework/src/Illuminate/Routing/Router.php(625): Illuminate\\\\Routing\\\\Router->dispatchToRoute()
#16 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(166): Illuminate\\\\Routing\\\\Router->dispatch()
#17 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\\\Foundation\\\\Http\\\\Kernel->Illuminate\\\\Foundation\\\\Http\\\\{closure}()
#18 /var/www/vendor/laravel/nova/src/Http/Middleware/ServeNova.php(24): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#19 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Laravel\\\\Nova\\\\Http\\\\Middleware\\\\ServeNova->handle()
#20 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#21 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\\\Foundation\\\\Http\\\\Middleware\\\\TransformsRequest->handle()
#22 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\\\Foundation\\\\Http\\\\Middleware\\\\ConvertEmptyStringsToNull->handle()
#23 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#24 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\\\Foundation\\\\Http\\\\Middleware\\\\TransformsRequest->handle()
#25 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\\\Foundation\\\\Http\\\\Middleware\\\\TrimStrings->handle()
#26 /var/www/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\\\Pipeline\\\\Pipeline->Illuminate\\\\Pipeline\\\\{closure}()
#27 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(141): Illuminate\\\\Pipeline\\\\Pipeline->then()
#28 /var/www/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\\\Foundation\\\\Http\\\\Kernel->sendRequestThroughRouter()
#29 /var/www/public/index.php(60): Illuminate\\\\Foundation\\\\Http\\\\Kernel->handle()
#30 {main}
```

**Comment by lisotton** [comment]
Just for you guys have an idea, I had this 18 times in the last 7 days:
![image](https://user-images.githubusercontent.com/1653860/129257465-eee06261-4d93-4465-8221-ce4006e8fec1.png)

**Comment by Pondake** [comment]
This issue still exists. When load-testing my app, it gives this error 5/250 times.

**Comment by swichers** [comment]
I am not sure how you would write a test for this without adding a parallel execution library.

The problem is that there is a race condition that happens when two requests deal with the same cache data and one clears it while the other is in the process of verifying and then reading it.

Let's say we have two requests, Request A and Request B. Request A comes in and winds up clearing the cache. Request B is requesting the cached data.

Request A -> Insert
Request B -> Check if cached data is available
Request A -> Clear cache
Request B -> Attempt to read cached data
Request B -> Throws exception

```php
// In CacheResponse handle
// There likely exists a race condition between hasBeenCached and getCachedResponseFor
// Code from v6.6.9

if ($this->responseCache->enabled($request)) {

  if ($this->responseCache->hasBeenCached($request, $tags)) {
    event(new ResponseCacheHit($request));

    // Request A (likely) nukes cache at this point

    $response = $this->responseCache->getCachedResponseFor($request, $tags);

    $this->getReplacers()->each(function (Replacer $replacer) use ($response) {
      $replacer->replaceInCachedResponse($response);
    });

    return $response;
  }
}
```

This is my guess anyway. After adding a cache clear to a specific endpoint on a very high traffic service, this error started appearing in large numbers. This service has a < 80ms response time with a few thousand users a minute, so I reckon we are hitting this edge case a lot more than is typical.

My first thought, if I am correct about the race condition, is to remove the call to hasBeenCached and just rely on checking the result from getCachedResponseFor.

edit: Maybe a better solution is to catch exceptions for that call, log, and then treat the request as uncached instead.

**Comment by lloricode** [comment]
> I am not sure how you would write a test for this without adding a parallel execution library.
> 
> The problem is that there is a race condition that happens when two requests deal with the same cache data and one clears it while the other is in the process of verifying and then reading it.
> 
> Let's say we have two requests, Request A and Request B. Request A comes in and winds up clearing the cache. Request B is requesting the cached data.
> 
> Request A -> Insert Request B -> Check if cached data is available Request A -> Clear cache Request B -> Attempt to read cached data Request B -> Throws exception
> 
> ```
> // In CacheResponse handle
> // There likely exists a race condition between hasBeenCached and getCachedResponseFor
> // Code from v6.6.9
> 
> if ($this->responseCache->enabled($request)) {
> 
>   if ($this->responseCache->hasBeenCached($request, $tags)) {
>     event(new ResponseCacheHit($request));
> 
>     // Request A (likely) nukes cache at this point
> 
>     $response = $this->responseCache->getCachedResponseFor($request, $tags);
> 
>     $this->getReplacers()->each(function (Replacer $replacer) use ($response) {
>       $replacer->replaceInCachedResponse($response);
>     });
> 
>     return $response;
>   }
> }
> ```
> 
> This is my guess anyway. After adding a cache clear to a specific endpoint on a very high traffic service, this error started appearing in large numbers. This service has a < 80ms response time with a few thousand users a minute, so I reckon we are hitting this edge case a lot more than is typical.
> 
> My first thought, if I am correct about the race condition, is to remove the call to hasBeenCached and just rely on checking the result from getCachedResponseFor.
> 
> edit: Maybe a better solution is to catch exceptions for that call, log, and then treat the request as uncached instead.

This is correct, we have a lot of project using this awesome library, and only one project occur this, and that project is a huge request, notice this when one of model resource is being updated and clearing cache, it will throw consecutive error of `unserialize`

**Comment by lloricode** [comment]
@swichers can you try PR of your idea?

**Comment by gmatkowski** [comment]
This is a huge issue for a projects that are big enough for this "race condition". 
In my personal opinion we should get rid of this "hasBeenCached" method and check a real value that exists/doesnt exist in "getCachedResponseFor" method.

**Comment by ahmadabd** [comment]
Upgrade to last version, it has been fixed in #383

**Comment by lisotton** [comment]
I'm not sure if #383 has fixed the issue. The exception won't be raised anymore, but the `hasBeenCached` will return `true` and `get` will return empty if there is a race condition. Or am I missing something reading the code of the PR?

**Reply by ahmadabd** [reply]
We can solve it by overriding handle method of CacheResoponse middleware.
Create new middleware
```
use Spatie\ResponseCache\Middlewares\CacheResponse;
class CacheResponseMiddleware extends CacheResponse
{
    public function handle(Request $request, Closure $next, ...$args): Response
    {
        $lifetimeInSeconds = $this->getLifetime($args);
        $tags = $this->getTags($args);

        if ($this->responseCache->enabled($request) && $this->responseCache->hasBeenCached($request, $tags)) {
            try {
                event(new ResponseCacheHit($request));

                $response = $this->responseCache->getCachedResponseFor($request, $tags);

                $this->getReplacers()->each(function (Replacer $replacer) use ($response) {
                    $replacer->replaceInCachedResponse($response);
                });

                return $response;
            } catch (CouldNotUnserialize $exception) {
                \Log::info("CacheResponse Race Condition exception accoured !!!");
                \Log::info($exception->getMessage());
            }
        }

        $response = $next($request);

        if ($this->responseCache->enabled($request) && $this->responseCache->shouldCache($request, $response)) {
            $this->makeReplacementsAndCacheResponse($request, $response, $lifetimeInSeconds, $tags);
        }

        event(new CacheMissed($request));

        return $response;
    }
}
```
Do nothing for new exception and cache again response.
just replace it in kernel.php with spatie CacheResponse middleware.

**Comment by swichers** [comment]
Like @lisotton said, that PR does not fix the problem being reported on this issue. It changes the nature of the failure. The race condition still exists, it's just been masked.

Replacing the middleware with your own is also not an ideal workaround because it duplicates the code into your own codebase. This makes you responsible for integrating future bugfixes or updates. If the module gets refactored your copied code may fail to continue to work. Not very DRY either.

The most likely proper fix for this is to rely on the response from `getCachedResponseFor` to know if something was actually cached or not. `hasBeenCached` should only be used when that's _all_ a request is checking.

Another option would be to refactor the method so several of its responsibilities are broken out into separate methods (more SOLID approach). At that point a custom middleware in your own codebase would not be duplicating a lot of third party package logic.

**Reply by ahmadabd** [reply]
Im agree with you about refactoring the method, it can help to solve the problem more efficient.
`hasBeenCached` helps to have better performane when cache does not exists, i think using exception to discover the Race Condition is good soloution to solve this problem. 
As you said refactoring the method can reduce side effect of making custom middleware.

**Reply by swichers** [reply]
The `hasBeenCached` call is technically slower since it is also doing a retrieval. Even if it wasn't a full retrieval, and was only a key lookup, it would still be technically slower because a failed cache retrieval already does a key lookup on the cache storage layer.

A key lookup vs a failed get should be exactly the same speed. If the result of your successful key lookup is always you retrieving the full item, you should just attempt to retrieve the full item to begin with.

But we don't have to take my word for it, let's look at the actual code:

hasBeenCached: `$this->taggedCache($tags)->has($this->hasher->getHashFor($request))`

getCachedResponseFor: `$this->taggedCache($tags)->get($this->hasher->getHashFor($request));`

They ultimately pass off logic to `ResponseCacheRepository` has/get which ultimately passes off to `Illuminate\Cache\Repository` has/get

Let's see how core implements the `has` call:

https://github.com/illuminate/cache/blob/master/Repository.php#L68=

```php
    public function has($key): bool
    {
        return ! is_null($this->get($key));
    }
```

So we can see the current code tries to retrieve the cached value twice when cache exists.

**Reply by ahmadabd** [reply]
Yes, you are right.

**Comment by roberttolton** [comment]
I'd like to +1 this issue, it's rare but annoying and difficult to explain to the unlucky few who might hit it when using the API.

Could we not just do (with some editing to `getCachedResponseFor` to have it return null if disabled / has no cache):

```
if ($this->responseCache->enabled($request) && $response = $this->responseCache->getCachedResponseFor($request, $tags)) {
  // use $response from now on?
}
```

Am I missing something?

Or in the interim, could we have it return an uncached response instead of just throwing the exception to the user?

**Comment by chris-lee-lb** [comment]
For this race condition problem. Can we refer to the nginx proxy cache lock mechanism for the same design ?

**Comment by swichers** [comment]
I have a pull request #434 which resolves this issue. I left in the technically unnecessary call to `hasBeenCached` to maintain closer alignment with the intent of the original code. It's possible someone could implement a cache provider where the `has()` call isn't equally as performance intensive has the `get()` call.
